internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -399,26 +400,28 @@ public void close(boolean gracefulClose) {
       if (isClosed) {
         return;
       }
-      long startTime = System.currentTimeMillis();
-      logger.info("Closing VeniceWriter for topic: {}", topicName);
-      try {
-        // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
-        // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
-        if (gracefulClose) {
-          endAllSegments(true);
-        }
-        // DO NOT call the {@link #PubSubProducerAdapter.close(int) version from here.}
-        // For non-shared producer mode gracefulClose will flush the producer
 
-        producerAdapter.close(topicName, closeTimeOutInMs, gracefulClose);
-        OPEN_VENICE_WRITER_COUNT.decrementAndGet();
-      } catch (Exception e) {
-        logger.warn("Swallowed an exception while trying to close the VeniceWriter for topic: {}", topicName, e);
-        VENICE_WRITER_CLOSE_FAILED_COUNT.incrementAndGet();
+      logger.info("Closing VeniceWriter for topic: {}, gracefulness: {}", topicName, gracefulClose);
+      try (Timer ignore = Timer.run(
+          elapsedTimeInMs -> logger.info("Closed VeniceWriter for topic: {} in {} ms", topicName, elapsedTimeInMs))) {
+        try {
+          // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
+          // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
+          if (gracefulClose) {
+            endAllSegments(true);
+          }
+          // DO NOT call the {@link #PubSubProducerAdapter.close(int) version from here.}
+          // For non-shared producer mode gracefulClose will flush the producer
+
+          producerAdapter.close(topicName, closeTimeOutInMs, gracefulClose);
+          OPEN_VENICE_WRITER_COUNT.decrementAndGet();
+        } catch (Exception e) {
+          handleExceptionInClose(e, gracefulClose);
```

Comment:
Originally, I was thinking whether we should implement this logic in the app layer of `VeniceWriter` since the app can decide whether it would choose to close it forcibly when some exception happens.
Maybe we should expose a param in case some usage of `VeniceWriter#close` would like to graceful close all the time and fail fast if it doesn't work.

Reply:
I thought about this idea initially and then realized that:
1. `close` has multiple callers (5-6), we could have some duplicated codes, if we handle it at its caller side.
2. More important, today's `close` function swallows all exceptions and doesn't pass them to its caller, so that it leaves the impression to its caller that `close` should always succeed (but actually it is not true). 

> Originally, I was thinking whether we should implement this logic in the app layer of VeniceWriter since the app can decide whether it would choose to close it forcibly when some exception happens.

Yeah, we can add another boolean in the signature and based on its value to determine if a forceful close is required. 
> Maybe we should expose a param in case some usage of VeniceWriter#close would like to graceful close all the time and fail fast if it doesn't work.

Reply:
What I meant is still keeping this logic in `VeniceWriter` to avoid duplication, but exposing a boolean param to let the user have an option to fallback or not.

Reply:
I think I did that in the new update. Would you take another look and see if it addressed your comments?